keps/sig-architecture/3784-database-of-names/README.md

```diff
@@ -0,0 +1,441 @@
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Kubernetes has many extensible fields which can carry either "standard"
+(contextual) or user-defined values.  Examples include (but are not limited to)
+label keys, annotation keys, node taints, pod resource names, conditions, etc.
+
+In most cases we define a "reserved" block, which is set aside for the project
+to use.  Labels and annotations which are prefixed with a `k8s.io` or
+`kubernetes.io` domain are reserved, and should not be used by end-users,
+whereas unprefixed names are reserved for end-users and should not be used by
+Kubernetes or any other project.
+
+We have made great use of these reserved names in myriad ways, all over the
+project.  What we have not had is a common way to know WHICH names we are
+using, what they mean, where they apply, or who is responsible for them.  In
+fact, the only way to find them is to read human-oriented documentation or to
+grep the codebase(s).
+
+This KEP proposes to store these names in a way that is searchable, machine
+parseable, but also relatively human-friendly.
+
+## Motivation
+
+We have too many KEPs and PRs that add annotation- or label-keys without any
+sort of collision detection or governance or documentation.
```

Comment:
and this solve it by creating some kind of git-database based on generating the secondary indexes?

Reply:
We can at least grow a culture of "hey, you should register that name"

Reply:
Yeah, I'm all in on that, but I think the key is to make it easy ... Having to PR another repo is always painful, at least to me ... How can make this simpler?

Reply:
Yeah, I don't know how to do that except putting it in k/k/staging, which I would be fine with.

Reply:
What if this would be autogenerated based on special from of godoc?

It would require additional machinery, but with the existing stuff in codegen maybe it's not ridiculously big thing to do.

Reply:
That could be - it seems like the *first* step is to define a place to store the results and a schema.  Will add this idea to the KEP.

Reply:
Added a note about staging, too